lib/asciidoctor/syntax_highlighter.rb

```diff
@@ -0,0 +1,82 @@
+module Asciidoctor
+  module SyntaxHighlighter
+    PROVIDED = {
+      'coderay' => 'syntax_highlighter/coderay',
+      'html-pipeline' => 'syntax_highlighter/html_pipeline',
+      'prettify' => 'syntax_highlighter/prettify',
+      'pygments' => 'syntax_highlighter/pygments',
+    }
+
+    attr_reader :name
+
+    def initialize doc, opts = {}; end
+
+    def docinfo? location; end
+
+    def highlight?; end
```

Comment:
Maybe we can avoid this method and just check for `highlight` ?
In other words, I don't think it makes sense to have `highlight? true` without an `highlight` method or `highlight? false` with an `highlight` method, right ?

Pushing this idea a bit further, I don't think we shouldn't allow the user to build "invalid object". I like the flexibility of Ruby but in this case we should maybe define a "trait" (replace with the correct Ruby terminology) for server-side highlighter since the following methods don't make sense in a client-side highlighter:

 *  `highlight`
 * `write_stylesheet?`
 * `library_available?`
 * `write_stylesheet`

Reply:
The Ruby way is to make "like a duck" objects. I tend to prefer that informal style over formal interfaces. This is consistent with how we did the extensions too.

What we're really lacking is documentation about how to work with the extension points in Asciidoctor. Once we have that, I think it becomes much more clear.

Reply:
I can see combining `highlight?` and `library_available?`. There's really no reason to claim to be able to highlight if the library isn't available, and it seems like it exposes too much detail. Either it can highlight or it can't.

Reply:
Actually, write_stylesheet? / write_stylesheet could be used by a client-side highlighter. Just because the highlighting happens on the client doesn't mean the stylesheet will automatically be there. That's up to the highlighter implementation.

So I think if we address highlight? / library_available? we end up with a unified API (except for the `highlight` function, which is understandable).

Reply:
> Actually, write_stylesheet? / write_stylesheet could be used by a client-side highlighter. Just because the highlighting happens on the client doesn't mean the stylesheet will automatically be there. That's up to the highlighter implementation.

Indeed you're right.

> So I think if we consolidate highlight? / library_available? we end up with a unified API (except for the highlight function, which is understandable).

:+1:

Reply:
:+1: